### PR TITLE
Tracking for propositions + split tracking event from record history event

### PR DESCRIPTION
**Pull request by spoorthipujariadobe**
## Description

1. Proposition tracking APIs implementation
2. Item level tracking APIs implementation
3. Split tracking event and separate it from event to record history for IAM
4. Renamed `MessagingPropositionItem` -> `PropositionItem`, `MessagingProposition` -> `Proposition` inline with iOS

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingFullscreenMessageDelegate.java [hunk 1]
```diff
@@ -61,8 +63,9 @@ public void onShow(final FullscreenMessage fullscreenMessage) {
     @Override
     public void onDismiss(final FullscreenMessage fullscreenMessage) {
         final InternalMessage message = (InternalMessage) fullscreenMessage.getParent();
-        if (message != null && message.getAutoTrack()) {
-            message.track(null, MessagingEdgeEventType.IN_APP_DISMISS);
+        if (message != null) {
+            message.recordEventHistory(null, MessagingEdgeEventType.DISMISS);
+            message.dismiss(false);
```
Comment (PravinPK) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingFullscreenMessageDelegate.java:68]: Should dismiss be called here?
I was under the assumption that onDismiss is called after the message has been dismissed. Wouldnt this recursively call message.dismiss()

File: code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingFullscreenMessageDelegate.java [hunk 2]
```diff
@@ -61,8 +63,9 @@ public void onShow(final FullscreenMessage fullscreenMessage) {
     @Override
     public void onDismiss(final FullscreenMessage fullscreenMessage) {
         final InternalMessage message = (InternalMessage) fullscreenMessage.getParent();
-        if (message != null && message.getAutoTrack()) {
-            message.track(null, MessagingEdgeEventType.IN_APP_DISMISS);
+        if (message != null) {
+            message.recordEventHistory(null, MessagingEdgeEventType.DISMISS);
```
Comment (PravinPK) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingFullscreenMessageDelegate.java:67]: In the dismiss method here we don't track the interaction. The tracking is happening at the InternalMessageClass.

Is it possible to refactor and keep tracking and recording to EventHistory in the same class?
Or if not can we add some comments mentioning tracking happens inside Message.dismiss()

**Comment by PravinPK** [comment]
Overall looks great. Few small questions.

File: code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/PropositionInteraction.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+// `PropositionInteraction` is a container for tracking related information needed to dispatch a
+// `decisioning.propositionDisplay` or `decisioning.propositionInteract` event to the Experience Edge.
+class PropositionInteraction {
+
+    private static final String SELF_TAG = "MessagingPropositionInteraction";
```
Comment (sbenedicadb) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/PropositionInteraction.java:21]: should we just have this be `PropositionInteraction` to match the class name?

**Review by sbenedicadb — APPROVED** [review]
looks good, couple nits

**Pull request merged**